mutex.go

```diff
@@ -45,14 +45,15 @@ var (
 func Lock(l sync.Locker, d time.Duration) error {
 	done := make(chan struct{}, 1)
 	m, ok := l.(*mutex)
+	var err error = nil
 	// if l is a mutex, lock with a timeout context.
 	if ok {
-		ctx, cancel := context.WithTimeout(context.Background(), d*time.Second)
+		ctx, cancel := context.WithTimeout(context.Background(), d)
 		defer cancel()
-		go func(context.Context) {
-			m.lock(ctx)
+		go func() {
+			err = m.lock(ctx)
 			done <- struct{}{}
-		}(ctx)
+		}()
 	} else {
 		go func() {
 			l.Lock()
```

Comment:
This occurred to me too, when I added the explicit check for the mutex type, why am I still trying to match the locker interface? I think it's good to support that interface, so that my class can be used as a replacement/enhancement for a mutex.Locker, but whereever I extend those semantics, it probably makes sense to explicity require a mutex as the passed arg, instead of acting like I can do something i can't with a base Locker.

So, I'm inclined to change the signature of mutex.Lock() to require a *mutex and remove the degraded (and buggy) support for passing the Locker interface on a non-mutex object. Does that make sense? Ditto for mutex.Unlock().